scrapy/core/http2/protocol.py

```diff
@@ -0,0 +1,183 @@
+
+        return d
+
+    def connectionMade(self):
+        """Called by Twisted when the connection is established. We can start
+        sending some data now: we should open with the connection preamble.
+        """
+        self.destination = self.transport.connector.getDestination()
+        LOGGER.info('Connection made to {}'.format(self.destination))
+
+        self.conn.initiate_connection()
+        self._write_to_transport()
+        self.is_connection_made = True
+
+        self._send_pending_requests()
+
+    def dataReceived(self, data):
+        events = self.conn.receive_data(data)
+        self._handle_events(events)
+        self._write_to_transport()
+
+    def connectionLost(self, reason=connectionDone):
+        """Called by Twisted when the transport connection is lost.
+        No need to write anything to transport here.
+        """
+        # Pop all streams which were pending and were not yet started
+        for stream_id in list(self.streams):
+            self.streams[stream_id].lost_connection()
+
+        self.conn.close_connection()
+
+        LOGGER.info("Connection lost with reason " + str(reason))
+
+    def _handle_events(self, events):
+        """Private method which acts as a bridge between the events
+        received from the HTTP/2 data and IH2EventsHandler
+
+        Arguments:
+            events {list} -- A list of events that the remote peer
+                triggered by sending data
+        """
+        for event in events:
+            LOGGER.debug(event)
+            if isinstance(event, ConnectionTerminated):
+                self.connection_terminated(event)
+            elif isinstance(event, DataReceived):
+                self.data_received(event)
+            elif isinstance(event, ResponseReceived):
+                self.response_received(event)
+            elif isinstance(event, StreamEnded):
+                self.stream_ended(event)
+            elif isinstance(event, StreamReset):
+                self.stream_reset(event)
+            elif isinstance(event, WindowUpdated):
+                self.window_updated(event)
+            else:
+                LOGGER.info("Received unhandled event {}".format(event))
+
+    # Event handler functions starts here
+    def connection_terminated(self, event: ConnectionTerminated):
+        pass
+
+    def data_received(self, event: DataReceived):
+        stream_id = event.stream_id
+        self.streams[stream_id].receive_data(event.data, event.flow_controlled_length)
+
+    def response_received(self, event: ResponseReceived):
+        stream_id = event.stream_id
+        self.streams[stream_id].receive_headers(event.headers)
+
+    def stream_ended(self, event: StreamEnded):
+        stream_id = event.stream_id
+        self.streams[stream_id].end_stream()
+
+    def stream_reset(self, event: StreamReset):
+        # TODO: event.stream_id was abruptly closed
+        #  Q. What should be the response? (Failure/Partial/???)
```

Comment:
In HTTP 1.1 the behavior [depends on the `DOWNLOAD_FAIL_ON_DATALOSS` setting](https://github.com/scrapy/scrapy/blob/master/scrapy/core/downloader/handlers/http11.py#L554).